bpo-33479: Tkinter docs

roseman opened the pull request:
bpo-33479: document threading issues in tkinter, other minor tkinter doc updates

native-api committed:
* Reduce the index page to a summary like the other ones

native-api committed:
* move seealso to the bottom
* insistent terminology: Tcl/Tk rather than Tk
* https://bugs.python.org/issue33479#msg316492 0. Intro: +supported Tcl/Tk versions and the two flavors; the fact that Tkinter adds its own logic, and the Tkinter doc concentrates on that logic

native-api committed:
more 0. intro

native-api committed:
https://bugs.python.org/issue33479#msg316492 1. Architecture.

native-api committed:
1. Architecture -- Tkinter modules

native-api committed:
2. Threading model

native-api committed:
start documenting contents

native-api committed:
* queue is per-thread

roseman committed:
update Tcl/Tk resource URLs and verbiage

roseman committed:
small rewording/reorganization of architecture and modules sections

roseman committed:
rewrite threading model section to separate normal use from potential traps

native-api commented on Doc/library/tkinter.rst:
```diff
@@ -11,100 +11,267 @@
-platforms, as well as on Windows systems.  (Tk itself is not part of Python; it
-is maintained at ActiveState.)
+the Tcl/Tk GUI toolkit. Tcl/Tk and :mod:`tkinter` are available on most Unix
+platforms (including macOS), as well as on Windows systems.
```
MacOS comment is redundant: elsewhere in the docs, e.g. https://docs.python.org/3/library/sys.html and https://docs.python.org/3/library/os.html , mentioning "Unix" implies MacOS, too; it's only mentioned separately when it differs from other UNIX flavors.

roseman replied:
my thought was since Mac vs. X11 were very different... but fair enough, especially at this point in the docs.

native-api replied:
My primary concern is this will prompt others to expect and add remarks like this elsewhere in the docs.

native-api replied:
https://github.com/roseman/cpython/pull/1

native-api commented on Doc/library/tkinter.rst:
```diff
@@ -11,100 +11,267 @@
+Tkinter is not a thin wrapper, but adds a fair amount of its own logic to
+make the experience more pythonic. This documentation will concentrate on these
+additions and changes, and refer to the official Tcl/Tk documentation for
+descriptions of underlying features.
```
The change to "underlying features" is factually incorrect. This is rather supposed to mean that outside custom features, Tkinter's interface has exactly the same semantic as Tk's one -- so references would not describe "underlying" features but Tkinter's features.

I see doc entries for such functions containing text like this: "Delegates to <Tcl/Tk function and link to it>."

roseman replied:
Does "its (or Tcl/Tk's) underlying features" address this?

zware replied:
Suggestion:

Tkinter adds significant logic to make the interface more Pythonic, but delegates most functionality to the underlying library.  This document defers to the official Tcl/Tk documentation for anything not added by Tkinter.

native-api replied:
I don't see what the problem was with my formulation. Suggested it @ https://github.com/roseman/cpython/pull/2

native-api commented on Doc/library/tkinter.rst:
```diff
@@ -11,100 +11,267 @@
+Architecture
+------------
+
+Tkinter provides a unified interface to several separate and distinct libraries.
```
* This phrase is meaningless.
* The fact that Tcl/Tk is a bundle rather that solid is critical to know to find your way around both Tkinter and Tcl/Tk doumentation -- without knowing that, its structure would make no sense to you. It thus can't be left out. It's also very unusual and surprising (for me, this was nothing short of shocking), that's why it's worth a special mention.

roseman replied:
It's not necessarily a bundle. The source versions are released as two separate tarballs, from two separate repositories. Many Linux distros distribute them as separate rpm's, meaning sometimes you get Tcl installed on a system but not Tk (usually when Tcl is pulled in as a dependency of some other package). Surprised me the first time I ran into it actually. Other times, e.g. with ActiveState, it is a single bundle. Your point is worth addressing, but I'm going to suggest doing it a bit later when it mentions the Tk library.

As a historical point of interest, distributing it as a strict core plus extensions available separately was another one of Ousterhout's "things", though many of us were pushing for a "batteries included" approach as Python uses. 

I also fail to see why you refer to this as "meaningless" as an intro sentence to an architecture section. But whatever.

zware replied:
I'm fine with what's here.

native-api replied:
It doesn't matter how they happen to be packaged. From architecture and documentation POV, they are separate.
(I don't really see a harm in this design: last I checked, the separation of concerns principle still existed. It's just that interested parties need to know this fact, that's all.)
The fact must be in the header of "Architecture" section because it justifies the section's existence. For most other bindings, just describing the interface is enough.
"several separate and distinct libraries" doesn't mean anything specific and only confuses the reader: "they said it's an interface to Tcl/Tk -- and now, it's some unspecified "several separate libraries"? The 1st part duplicates the intro. So, the sentence has negative informativity.

native-api replied:
https://github.com/roseman/cpython/pull/3

native-api commented on Doc/library/tkinter.rst:
```diff
@@ -11,100 +11,267 @@
+
+Tcl
+   Like Python, Tcl is a dynamic interpreted programming language. It is most
+   commonly used by C applications to embed a scripting language, or as an
```
not "or" but "and". Tk can only be used in combination with Tcl, whether the app needs scripting or not.

roseman replied:
There's a lovely ambiguity in the English language around common use of "or" and "and" which also drives everyone with a database or formal logic background crazy. Whichever works for you is fine.

As for Tk without Tcl, ask the Perl guys, but except for that horrible mistake, your statement is true. However, Tcl is very frequently used without Tk, which was what I was trying to say there.

zware replied:
English is horrible (saying this as a monolingual speaker), but the "or" is fine.

native-api replied:
You see that I was already confused -- and I already know the stuff! So, let's get around this "lovely ambiguity", if only to heed Python Zen.

native-api replied:
https://github.com/roseman/cpython/pull/4

native-api commented on Doc/library/tkinter.rst:
```diff
@@ -11,100 +11,267 @@
+   create and manage one or more instances of a Tcl interpreter, run Tcl
+   commands and scripts in those instances, and add custom commands
+   implemented in either Tcl or C. Tcl provides for an event queue, used
+   for I/O, timers, and by modules.
```
"used for etc" is vague (so best left out as useless for the reader) and omits the critical fact that it's implemented by and tied to the interpreter itself (i.e. I can't create separate instances of the queue).

roseman replied:
This one I disagree with. The examples seem concrete to me, and by saying that it's available for multiple different things including for modules gives us info about how Tk hooks into it. This was actually following your lead a bit in that most people tend to associate event queues and event loops with a GUI toolkit, so the fact that it's part of the underlying language is unexpected. It actually used to be in Tk rather than Tcl, but that was a while back.

Regarding the omission, is it that you're reading it as the event queue is separate from the interpreter? Would saying "The Tcl interpreter maintains an event queue..." be clearer? The separate instances thing I don't quite get.

native-api replied:
The problem is, this is not useful information for a Tkinter user, and is confusing:
"What? How in the world can an event queue be used for I/O (it's not like the OS is listening to it, right?), or timers? (looks for an explanation further on, and doesn't find it)".
What "used by modules" means is a mystery to me. The general meaning of "module" is "some separate code chunk". Then it's basically saying "it can be used by code", which is a remark worthy of Captain Obvious.

> Regarding the omission

No, I read it as "doesn't say about the relationship at all".

In other languages, queues are just another data structures that can be created and used as one sees fit. This "Tcl queue" is not like them, it's special, and the text fails to address this.

Look at https://github.com/native-api/cpython/blob/6cdcb9e2858ab8b2417ebcd771dd0ef9361b8878/Doc/library/tkinter.rst for my idea how to put the relationship.

roseman replied:
How about "Tcl provides an extensible framework for event-driven programming, used by Tk for user interface events." and then delete the part in the Tk section?

native-api replied:
... the 2nd comment @ https://stackoverflow.com/a/185789/648265 %)

native-api replied:
https://github.com/roseman/cpython/pull/5

native-api commented on Doc/library/tkinter.rst:
```diff
@@ -11,100 +11,267 @@
+   for I/O, timers, and by modules.
+   
+Tk
+   Tk is a module that can be loaded into a Tcl interpreter instance. It adds
```
Vague, omits that it's a _Tcl_ module and adds _custom Tcl_ commands -- those very ones that were mentioned in the prev. paragraph. This text fails to make that connection.

roseman replied:
Agree. Thanks.

Question: Tcl uses the term "package" for what Python calls a module. Would it be better to say up front something to the effect of "the tcl interpreter can load packages to add features (written in C or Tcl, akin to Python modules), and then thereafter use "packages" when referring to Tcl add-ons? The alternative I think would be saying "Tcl module" everywhere vs. just "module" for Python.

native-api replied:
Use Tcl's official terminology to describe their relationship -- that way, the text will have an exact meaning.
No need to go into details: the reader only needs to know that: 1)it's some kind of loadable add-on; 2)it's implemented in C (which would explain how it can do GUI when Tcl itself can't); 3) it plugs itself into Tcl's machinery by adding custom Tcl commands.

native-api replied:
https://github.com/roseman/cpython/pull/6

native-api commented on Doc/library/tkinter.rst:
```diff
@@ -11,100 +11,267 @@
+   
+Tk
+   Tk is a module that can be loaded into a Tcl interpreter instance. It adds
+   Tcl commands (implemented in C) to create and manipulate GUI widgets. Each
```
Wrong, some are implemented in Tcl.

roseman replied:
Correct. I'd put that because the version I based this on had included it, though I think it's pretty much irrelevant either way. Happy if it were removed altogether.

native-api replied:
>Happy if it were removed altogether.

Agreed. Mentioning the fact that the _module_(/package/whatchamacallit) is implementeed in C already explains how it can do what it does.

native-api replied:
https://github.com/roseman/cpython/pull/6

native-api commented on Doc/library/tkinter.rst:
```diff
@@ -11,100 +11,267 @@
+   Tk is a module that can be loaded into a Tcl interpreter instance. It adds
+   Tcl commands (implemented in C) to create and manipulate GUI widgets. Each
+   :class:`Tk` object embeds its own Tcl interpreter instance with Tk loaded into
+   it. Tk's widgets are very customizable, though at the cost of a dated appearance.
```
A value judgement.

roseman replied:
Would it be better to say "Tk's widgets were designed to support extensive visual customizations which do not map well to modern user interface conventions on various platforms."

To explain further, this over-the-top customization is one of the reasons the whole ttk thing is a separate set of widgets rather than just making the classic ones look right, which would break all the customizations. (In practice they did both, i.e. some of the classic widgets were changed to use platform api's which broke some of the customization.). But the reason for including the statement is that it puts ttk in context.

Given this documentation is far from a complete reference, I think it's already having an opinion (what to include and what not to include). Differentiating between the two widget sets seems as much a "reference" thing as anything else.

native-api replied:
Mentioning in Ttk's paragraph, concisely, that it's intended as a replacement and the key official points how it's supposed to be better is enough IMO. To a programmer's ear, "intended as a more modern/whatever replacement" already says: "recommended for use in new code" (can as well say this explicitly if the official documentation does).

native-api replied:
https://github.com/roseman/cpython/pull/8

native-api commented on Doc/library/tkinter.rst:
```diff
@@ -11,100 +11,267 @@
+   it. Tk's widgets are very customizable, though at the cost of a dated appearance.
+   Tk uses Tcl's event queue to generate and process GUI events. Note
+   that unlike some GUI libraries, each interpreter uses only a single thread,
+   which has implications for :mod:`tkinter` users (see `Threading model`_).
```
Threading stuff is Tcl, not Tk, so is out of place here.

roseman replied:
I'd be happy just deleting that last sentence, which is just a forward reference to the threading thing anyway.

native-api replied:
Agreed.

native-api replied:
https://github.com/roseman/cpython/pull/7

native-api commented on Doc/library/tkinter.rst:
```diff
@@ -11,100 +11,267 @@
+   Themed Tk (Ttk) is a newer family of Tk widgets that provide a much better
+   appearance on different platforms than many of the classic Tk widgets.
+   Ttk is distributed as part of Tk, starting with Tk version 8.5. Python
+   bindings are provided in a separate module, :mod:`tkinter.ttk`.
```
Factually & structurally incorrect. Ttk is not a separate module, so can't have a separate entry. It's not "distributed" as part of Tk but is implemented by it an an integral part. ("Distributed" is vague, too -- ususally means separate files.)
Judgements, too.

roseman replied:
Would "Ttk is part of Tk, starting..." and "Python bindings are provided in the module tkinter.ttk" be a better way of saying this? 

Having some form of separation seems to me important, as tkinter.ttk isn't guaranteed to work everywhere tkinter is (or has support for tk 8.4 officially been dropped?)

native-api replied:
"Since version 8.5, Tk also implements Themed Tk (Ttk)..."

See https://github.com/python/cpython/pull/7287#discussion_r192571941 on the 2nd part.

Also see https://github.com/python/cpython/pull/7287#discussion_r192572833 about expressing its relationship to regular Tk.

native-api replied:
https://github.com/roseman/cpython/pull/8

native-api commented on Doc/library/tkinter.rst:
```diff
@@ -11,100 +11,267 @@
+   bindings are provided in a separate module, :mod:`tkinter.ttk`.
+
+Tix
+   `Tix`<https://core.tcl.tk/jenglish/gutter/packages/tix.html>_ is an older
```
Judgements.

zware replied:
Accurate.  Tix is abandoned (the last release was over 10 years ago) and deprecated, and will be removed as soon as we have any trouble building it for the Windows releases.

Frankly, I'd rather it out altogether here so we have fewer places to fix when we do remove it.

roseman replied:
Older is factually true. Python people are keeping it on life support but that's the extent of it. But as long as its noted as deprecated, fine.

native-api replied:
[As Joel Spolsky writes](https://www.joelonsoftware.com/2000/04/06/things-you-should-never-do-part-i/), _code doesn't rust._ So "older" and "abandoned" are moot points.

"Deprecated in favor of Ttk" is the precise formulation. In particular, "deprecated" also means that support can be dropped at any moment in the future.

It could say something about their relationship that would shed light on why it's considered obsolete without attaching labels -- e.g.: "Widgets provided by Tix have since been implemented in Ttk, and Tix is deprecated in favor of it." (if that is correct, of course).

native-api replied:
From http://tix.sourceforge.net/dist/current/man/html/contents.htm and https://www.tcl.tk/man/tcl8.5/TkCmd/contents.htm , it's not clear if all "useful" Tix widgets have Ttk replacements.

native-api replied:
> As Joel Spolsky writes, code doesn't rust. So "older" and "abandoned" are moot points.

I mean, it can't be deprecated _just_ because it's "old" or the original author lost interest (PyXML is a recent counterexample). It must be somehow inferior to other things to lose competition to them.

Anyway, https://github.com/roseman/cpython/pull/9

native-api commented on Doc/library/tkinter.rst:
```diff
@@ -11,100 +11,267 @@
+Most applications will directly use the features provided by :mod:`tkinter`.
+Unless compatibility with very old versions of Tcl/Tk is required, Ttk widgets
+found in :mod:`tkinter.ttk` should also be favoured over their classic Tk
+counterparts.
```
Advice is none of a reference doc's business, it's user's guide domain.
We already said that Ttk is intended by its authors as a newer replacement.

roseman replied:
I still disagree, given this reference documentation is no more than an overview, but if you think there's value in excluding this, it's fine with me.

native-api replied:
I can't see what was wring with my version. It said clearly that you will need `tkinter` only unless you're using additional widgets.

Oh, I've got it. The formulation should be neutral -- mention "additional" widgets on equal grounds with "base" ones. Ttk's section should already say how it is superior as per https://github.com/python/cpython/pull/7287#discussion_r192572243.

I'll suggest some.

native-api replied:
https://github.com/roseman/cpython/pull/8

native-api commented on Doc/library/tkinter.rst:
```diff
@@ -11,100 +11,267 @@
+application programmers. It may be needed to access features in very new versions
+of Tcl/Tk without an existing Python binding, though this reduces compatibility.
+It is usually a shared library (or DLL) but may be statically linked with the
+Python interpreter.
```
Likewise, usage advice. Ref doc's business is only to state that it's not public API except some parts (I'm going to claim `tkapp.call` at least as public, that was written in the plan).

native-api commented on Doc/library/tkinter.rst:
```diff
@@ -11,100 +11,267 @@
+    to events or threads, be aware of these possibilities.
+
+  * There are a few select :mod:`tkinter` functions that presently work only when
+    called from the thread that created the Tcl interpreter.
```
Long story short, the section is partly vague, partly incorrect, partly conceals things outright. Since literally no-one on the Net knows how the damn thing works, it's absolutely imperative to describe the user-visible semantics fully and precisely.

I agree with toning down the comparison with "other GUI toolkits" though.

I think that the two GUI execution models needs to be described this way or another still, because without knowing them, the other explanations and Tkinter's interface and usage models will not make any sense to the reader.
Tkinter reconciliates two very different usage pattenrs, literally two different worlds, that's a legitimately hard-to-grasp and confusing design element, it positively needs a quality explanation.

roseman replied:
This is going to come across sounding sarcastic or condescending, but it's actually not... I've never met someone who has run into such difficulties with this as you have. Which is incredibly useful when it comes to documentation, so I'm finding this whole exchange very interesting and helpful. Personally I'd like to know more of the details about what works and what doesn't too.

I'm also conscious that if we essentially have one piece of reference documentation to say what the module provides, and that contains only a subset of what is actually provided, I'd want what's in there to be relevant for most users looking at how to use the module. I'm honestly not sure how much of this material should be included by that standard.

native-api replied:
I'll provide a middle-ground version as a PR against your repo.

roseman replied:
Whatever is easiest moving forward. If you just want to integrate things into what you had previously that is more than cool with me. As you can tell, I'm just a little Github-rusty!

native-api replied:
https://github.com/roseman/cpython/pull/10

native-api reviewed (commented):
[A reference documentation's job is to fully, precisely and authoritatively describe the subject's public interface (which includes semantic) -- nothing more, nothing less.](https://meta.serverfault.com/questions/8934/what-to-do-with-questions-when-the-answer-is-in-a-man-page#comment22241_8938)

There is thus no place in it for vague statements (because they have zero to negative informativity), weasel words, value judgements, factual inaccuracies and omitting or deliberately concealing any details critical to understanding the software's operation.

This is all the more critical that no-one on the Net, including the Python team, seems to really understand how Tkinter works, and this heavily stifles its adoption.

Likewise, reference doc has no business telling the user what to do and what not to do -- "I can decide that for myself, thank you very much!" It solely tells what is supported and what is not.
Actually advising the users how to use the software is the domain of a user's guide (that may include a tutorial).

zware commented on Doc/library/tkinter.rst:
```diff
@@ -11,100 +11,267 @@
-
-   Tcl/Tk documentation:
+Tkinter supports a range of Tcl/Tk versions, built either with or
+without thread support. The official Python binary release bundles Tcl/Tk 8.6
```
"The Windows binary release provided by python.org bundles Tcl/Tk 8.6 with thread support."

native-api replied:
It's "binary release", precisely. I specifically checked that this stands for the MacOS release, too.

native-api replied:
Didn't say "binary release_s_" because that could be interpreted as referencing Python releases in general, i.e. other versions, too, which is untrue.

zware commented on Doc/library/tkinter.rst:
```diff
@@ -11,100 +11,267 @@
+Tkinter supports a range of Tcl/Tk versions, built either with or
+without thread support. The official Python binary release bundles Tcl/Tk 8.6
+threaded. See the source code for the :mode:`_tkinter` module
+for more information about supported versions.
```
Referring to the C source for a private module isn't going to cut it.  Let's just say "8.5 and above" and leave it at that.  Along with the above suggestion, that would make the whole paragraph:

```
Tkinter supports most builds of Tcl/Tk version 8.5 and above, including builds
with and without thread support.  The Windows and macOS binary releases
provided by python.org bundle Tcl/Tk 8.6 with thread support.
```

We'd need @ned-deily to confirm that's what the macOS builds provide, though.

native-api replied:
Adding these details to the docs will be problematic: we don't really know what versions we support, it's not documented anywhere. We only test it with the versions that binary releases are built with AFAIK. So claiming official support for the entire range is an outright lie which would most probably bite someone in the butt if they try to rely on it.

That's why I preferred to just gloss over this fact and only say what we truly know.  The source code is the best source of this kind of information we currently have. And while this can be discussed, documented, and buildbots set up or something, it's certaintly of no use to stall publishing of the docs until then.

If someone needs to build Tkinter against other version, they'll need its sources anyway, and searching them for "version" isn't too hard.

zware reviewed (changes requested):
I haven't made it all the way through this, but have added a few thoughts.

Thank you to both @native-api and @roseman for working on this; the tkinter docs have been in want of an update for quite some time.

native-api committed:
https://github.com/python/cpython/pull/7287#discussion_r192545213

native-api committed:
https://github.com/python/cpython/pull/7287#discussion_r192545445

roseman committed:
Merge pull request #1 from native-api/roseman_1

https://github.com/python/cpython/pull/7287#discussion_r192545213

roseman committed:
Merge pull request #2 from native-api/roseman_2

https://github.com/python/cpython/pull/7287#discussion_r192545445

native-api committed:
https://github.com/python/cpython/pull/7287#discussion_r192545697

native-api committed:
mention Tix as it adds more complexity to the architecture

native-api committed:
spec

native-api committed:
https://github.com/python/cpython/pull/7287#discussion_r192545744

native-api committed:
https://github.com/python/cpython/pull/7287#discussion_r192545829

native-api committed:
more specific

native-api committed:
https://github.com/python/cpython/pull/7287#discussion_r192545899
https://github.com/python/cpython/pull/7287#discussion_r192545929

native-api committed:
https://github.com/python/cpython/pull/7287#discussion_r192545996

native-api committed:
https://github.com/python/cpython/pull/7287#discussion_r192546347

roseman committed:
plural releases

roseman committed:
Merge pull request #3 from native-api/roseman_3

https://github.com/python/cpython/pull/7287#discussion_r192545697

roseman committed:
Merge pull request #4 from native-api/roseman_4

https://github.com/python/cpython/pull/7287#discussion_r192545744

roseman committed:
Merge pull request #5 from native-api/roseman_5

https://github.com/python/cpython/pull/7287#discussion_r192545829

roseman committed:
Merge pull request #6 from native-api/roseman_6

https://github.com/python/cpython/pull/7287#discussion_r192545899

roseman committed:
Merge pull request #7 from native-api/roseman_7

https://github.com/python/cpython/pull/7287#discussion_r192545996

roseman committed:
put back 'older' for tix

Despite the "code does not rust", UI certainly does with respect to visual integration into modern applications. While an API may stay functional, if it says "gimme a Widget X" and what comes up is some ugly piece of crap that doesn't fit in with the rest of the UI, it's not meeting the programmer's expectations. I'd prefer a flashing 'warning' light, a phrase like 'virtually abandoned', etc. but I'd settle for 'older'. Something needs to be there.

roseman committed:
Merge pull request #9 from native-api/roseman_9

https://github.com/python/cpython/pull/7287#discussion_r192546347
